### PR TITLE
fix rollup data missing points in rob

### PR DESCRIPTION
**Pull request by zlin179**
Fix for issue https://github.com/grafana/metrictank/issues/1485

**Describe your changes**
The rollup aggregators only get the points once they leave the ROB.
This PR fixs aggregator so when being queried it will return the current value (considering points in ROB) but will not actually consume the points in ROB.

**Testing performed**
* Add some unit tests for it and pass it.
* Integrate test by comparing raw data and rollup data queried at the same time

debug

### CODE REVIEW

**Comment by shanson7** [comment]
The images in the description are a little confusing. 

1. It seems like they should be reversed to match before/after (e.g. before the fix and then after the fix).
2. I think the mention of 31d should really just be `5-minute rollups` so it has more context outside of BB
3. The first image (the "after") has the wrong text "no 16:55" should probably be "includes 16:30" or "includes most recent time interval".

**Comment by shanson7** [comment]
Also in the future it would be good to do multiple commits like:

1. First commit - add test case (that fails)
2. Second commit - add the fix (test case should now pass)

Any other commits (like code cleanup or similar) should also be separate from those 2 commits.

File: mdata/aggmetric.go [hunk 1]
```diff
@@ -148,8 +148,8 @@ func (a *AggMetric) SyncAggregatedChunkSaveState(ts uint32, consolidator consoli
 
 func (a *AggMetric) GetAggregated(consolidator consolidation.Consolidator, aggSpan, from, to uint32) (Result, error) {
 	// no lock needed cause aggregators don't change at runtime
-	for _, a := range a.aggregators {
-		if a.span == aggSpan {
+	for _, aggregator := range a.aggregators {
```
Comment (shanson7) [inline mdata/aggmetric.go:151]: This rename change is good (removes shadowing of `a`) but should have been in a separate commit.

File: mdata/aggmetric.go [hunk 2]
```diff
@@ -163,25 +163,38 @@ func (a *AggMetric) GetAggregated(consolidator consolidation.Consolidator, aggSp
...
+				agg = aggregator.sumMetric
 			default:
 				err := fmt.Errorf("internal error: AggMetric.GetAggregated(): unknown consolidator %q", consolidator)
 				log.Errorf("AM: %s", err.Error())
 				badConsolidator.Inc()
 				return Result{}, err
 			}
 			if agg == nil {
-				return Result{}, fmt.Errorf("Consolidator %q not configured", consolidator)
+				return Result{}, fmt.Errorf("consolidator %q not configured", consolidator)
```
Comment (shanson7) [inline mdata/aggmetric.go:182]: Why change this line?

File: mdata/aggmetric.go [hunk 3]
```diff
@@ -163,25 +163,38 @@ func (a *AggMetric) GetAggregated(consolidator consolidation.Consolidator, aggSp
...
 			}
-			return agg.Get(from, to)
+			futurePoints := []schema.Point{}
+			if a.rob != nil {
+				futurePoints = a.rob.Get()
+			}
+			result.Points = aggregator.Foresee(consolidator, from, to, futurePoints)
+			if result.Oldest > result.Points[0].Ts {
+				result.Oldest = result.Points[0].Ts
+			}
```
Comment (shanson7) [inline mdata/aggmetric.go]: Couldn't all of this be in the `if` statement above? That reduces the changes required for the non-rob case.

File: mdata/aggmetric_test.go [hunk 4]
```diff
@@ -486,14 +486,13 @@ func TestGetAggregated(t *testing.T) {
 	m.Add(30, 30)
 	m.Add(31, 31)
 	m.Add(32, 32)
-	m.Add(40, 40)
```
Comment (shanson7) [inline mdata/aggmetric_test.go:489]: Why remove this?

File: mdata/aggregator.go [hunk 5]
```diff
@@ -99,6 +100,38 @@ func (agg *Aggregator) flush() {
...
+			if aggregationPreview.Cnt != 0 {
+				value, _ := aggregationPreview.GetValueFor(consolidator)
+				aggregatedPoints = append(aggregatedPoints, schema.Point{Val: value, Ts: currentBoundary})
+				aggregationPreview.Reset()
+			}
+			currentBoundary = boundary
+		}
+		aggregationPreview.Add(point.Val)
+	}
+	if currentBoundary <= to {
```
Comment (shanson7) [inline mdata/aggregator.go]: Should there be a check for `aggregationPreview.Cnt != 0` here?

File: mdata/aggregator_test.go [hunk 6]
```diff
@@ -198,25 +203,99 @@ func TestAggregator(t *testing.T) {
...
+	comparePoints(
+		"simple-sum-foresee",
+		agg.Foresee(consolidation.Sum, 0, 1000, futurePoints),
+		[]schema.Point{
+			{Val: 10, Ts: 300},
+			{Val: 4, Ts: 360},
+		},
+	)
+
+	// Foresea should not change the acutal state of aggregator
```
Comment (shanson7) [inline mdata/aggregator_test.go]:
```suggestion
	// Foresee should not change the actual state of aggregator
```

**Review by shanson7 — CHANGES_REQUESTED** [review]

File: mdata/aggmetric_test.go (hunk 4 above)
Comment (zlin179) [reply mdata/aggmetric_test.go:489]: Well there's no ROB in this case. 40 is the boundary of that span. It will return [40, 40] anyway even without `Foresee`. Here I am trying to test if it can return the boundary if the last point is somewhere in the middle and `to >= boundary`

File: mdata/aggmetric.go (hunk 2 above)
Comment (zlin179) [reply mdata/aggmetric.go:182]: My linter kept complaining that the first letter in `Errorf` should be lower-case. As it is the case for other `Errorf` in this file, I changed it.

Comment (shanson7) [reply mdata/aggmetric.go:182]: Reasonable, but again (in the future) do these things in a separate commit.

File: mdata/aggmetric_test.go (hunk 4 above)
Comment (shanson7) [reply mdata/aggmetric_test.go:489]: But maybe that was on purpose for testing behavior on the boundary condition? Perhaps your intended test should be an additional case?

Comment (zlin179) [reply mdata/aggmetric_test.go:489]: Well I think the boundary thing (specifically just a `if` branch) is part of the `aggmetric.GetAggregated`. We add some points to test some branches, add some other points to other branches. They are sequentially tested by `expected := []schema.Point`. Here to be convenient I just delete one point because I think 9 points for the previous test (the original logic) is enough. I can also add a new point for the new logic, but I think it's unnecessary to separate them into two functions as they test the same function?

**Comment by zlin179** [comment]
> Also in the future it would be good to do multiple commits like:
> 
> 1. First commit - add test case (that fails)
> 2. Second commit - add the fix (test case should now pass)
> 
> Any other commits (like code cleanup or similar) should also be separate from those 2 commits.

Got it. It's just that some repos require clean history and some repos don't care about it lol. I just read the CONTRIBUTING.md here
I kind of twisted the commit history here.

File: mdata/aggmetric.go (hunk 1 above)
Comment (zlin179) [reply mdata/aggmetric.go:151]: well, I renamed this because I need the access the outer `a` `a *AggMetric` somewhere in the forloop (specifically the `a.rob` below) while the original code doesn't not need to . If I separate it the code won't work in the first commit

**Comment by zlin179** [comment]
> The images in the description are a little confusing.
> 
> 1. It seems like they should be reversed to match before/after (e.g. before the fix and then after the fix).
> 2. I think the mention of 31d should really just be `5-minute rollups` so it has more context outside of BB
> 3. The first image (the "after") has the wrong text "no 16:55" should probably be "includes 16:30" or "includes most recent time interval".

Updated Image
mt-fix

**Comment by shanson7** [comment]
> > Also in the future it would be good to do multiple commits like:
> > 
> > 1. First commit - add test case (that fails)
> > 2. Second commit - add the fix (test case should now pass)
> > 
> > Any other commits (like code cleanup or similar) should also be separate from those 2 commits.
> 
> Got it. It's just that some repos require clean history and some repos don't care about it lol. I just read the CONTRIBUTING.md here I kind of twisted the commit history here.

I think the history is mildly important, but it's also nice when reviewing to decouple the incidental commits and the core commits. I can review the incidental stuff first and then just look at the core commits separately.

File: mdata/aggmetric.go (hunk 1 above)
Comment (shanson7) [reply mdata/aggmetric.go:151]: The first commit could be just the variable rename. No worries though

**Review by shanson7 — APPROVED** [review]
LGTM. I changed the base branch so this can be merged. Then we can create a PR from this repo to grafana/metrictank
